File: Complementar - Estruturas de dados (Didático)/Algoritmos17/Alg17.cpp
/*
♦ Algoritmo 17: Estrutura de Dados Linear Dinâmica (não Estática) de Classificação/Ordenação em tempo linear Counting Sort (Ordenação por Contagem).
*/

#include <iostream>
#include <vector>

using namespace std;

void countingSort(vector<int> &arr) {
    int max = *max_element(arr.begin(), arr.end());
    int min = *min_element(arr.begin(), arr.end());

    int range = max - min + 1;
    vector<int> countArray(range, 0);
    vector<int> outputArray(arr.size());

    // Inicializa o array de contagem
    for (int i = 0; i < arr.size(); i++) {
        countArray[arr[i] - min]++;
    }

    // Atualiza o array de contagem para ter a soma acumulativa
    for (int i = 1; i < range; i++) {
        countArray[i] += countArray[i - 1];
    }

    // Constrói o array de saída
    for (int i = arr.size() - 1; i >= 0; i--) {
        outputArray[countArray[arr[i] - min] - 1] = arr[i];
        countArray[arr[i] - min]--;
    }

    // Copia o array ordenado de volta para o array original
    for (int i = 0; i < arr.size(); i++) {
        arr[i] = outputArray[i];
    }
}

void printArray(const vector<int> &arr) {
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;
}

int main() {
    vector<int> arr = {4, 2, 2, 8, 3, 3, 1};
    cout << "Array original:" << endl;
    printArray(arr);

    countingSort(arr);

    cout << "Array ordenado:" << endl;
    printArray(arr);

    return 0;
}

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos17/Alg17.py
'''
♦ Algoritmo 17: Algoritmo de Classificação/Ordenação para Estruturas de Dados Lineares Dinâmicas (não Estáticas): Counting Sort (Ordenação por Contagem).
'''

def counting_sort(arr):
    max_value = max(arr)
    min_value = min(arr)

    range_size = max_value - min_value + 1
    count_array = [0] * range_size
    output_array = [0] * len(arr)

    # Inicializa o array de contagem
    for num in arr:
        count_array[num - min_value] += 1

    # Atualiza o array de contagem para ter a soma acumulativa
    for i in range(1, range_size):
        count_array[i] += count_array[i - 1]

    # Constrói o array de saída
    for i in range(len(arr) - 1, -1, -1):
        output_array[count_array[arr[i] - min_value] - 1] = arr[i]
        count_array[arr[i] - min_value] -= 1

    # Copia o array ordenado de volta para o array original
    for i in range(len(arr)):
        arr[i] = output_array[i]

def print_array(arr):
    print(' '.join(map(str, arr)))

arr = [4, 2, 2, 8, 3, 3, 1]
print("Array original:")
print_array(arr)

counting_sort(arr)

print("Array ordenado:")
print_array(arr)
